tools/codex_schema.py

```
@@ -3323,10 +3420,94 @@ def instance_valid(instance: Any, schema: Any, root: dict[str, Any], depth: int)
+    for method, root_names in shared.items():
+        names = (root_names,) if isinstance(root_names, str) else root_names
+        present = [name in roots for name in names]
+        if any(present) and not all(present):
+            fail(f"selected shared Codex method {method} has incomplete roots")
+        if all(present):
+            selected[method] = root_names
+    return selected
+
+
+def verify_required_field_rejections(
+    version: str, contract_name: str, instance: Any, schema: Any
```

Ruff (0.16.2)

[warning] Dynamically typed expressions (typing.Any) are disallowed in `instance`

(ANN401)